File: Solutions/Python/binary_tree_zigzag_level_order_traversal.py
"""
    [MEDIUM]
    103. Binary Tree Zigzag Level Order Traversal

    Concepts:
    - binary tree
    - breadth-first search

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 17.94 MB  [Beats 26.58%]
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        # level order, but reverse every other level
        levels = []
        direction = 1

        if not root:
            return []

        queue = deque([root])
        while queue:
            n = len(queue)
            level = []

            for _ in range(n):
                curr = queue.popleft()
                level.append(curr.val)

                if curr.left:
                    queue.append(curr.left)
                if curr.right:
                    queue.append(curr.right)

            levels.append(level[::direction])
            direction *= -1

        return levels

File: Solutions/Cpp/binary_tree_zigzag_level_order_traversal.cpp
/*
    [MEDIUM]
    103. Binary Tree Zigzag Level Order Traversal

    Concepts:
    - binary tree
    - breadth-first search

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 15.10 MB  [Beats 76.77%]
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        std::vector<std::vector<int>> levels = {};
        std::vector<int> level = {};

        if (!root)
            return levels;

        std::queue<TreeNode*> q;
        q.push(root);
        int n;
        bool is_forward = true;
        TreeNode* curr;

        while (!q.empty()) {
            n = q.size();
            level.clear();

            for (int _ = 0; _ < n; _++) {
                curr = q.front();
                q.pop();
                level.push_back(curr->val);

                if (curr->left)
                    q.push(curr->left);
                if (curr->right)
                    q.push(curr->right);
            }

            if (!is_forward)
                std::reverse(level.begin(), level.end());

            is_forward = !is_forward;
            levels.push_back(level);
        }

        return levels;
    }
};